src/truncate.rs

```diff
@@ -0,0 +1,210 @@
+
+use crate::errors::Error;
+use crate::errors::Error::InvalidArgument;
+use crate::file_pipe_log::{build_file_reader, build_file_writer, LogFileReader, LogFileWriter};
+use crate::file_pipe_log::{LogFd, ReplayMachine};
+use crate::log_batch::LogItemBatch;
+use crate::log_batch::LogItemContent::{Command, EntryIndexes, Kv};
+use crate::memtable::EntryIndex;
+use crate::truncate::TruncateMode::{All, Back, Front};
+use crate::{log_batch, FileBuilder, FileId, LogBatch, LogQueue};
+use std::collections::HashMap;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+use std::sync::Arc;
+
+pub struct TruncateMachine {
+    items: Vec<LogItemBatch>,
+}
+
+#[derive(Clone, Copy)]
+pub enum TruncateMode {
+    Front = 0,
+    Back = 1,
+    All = 2,
+}
+
+impl FromStr for TruncateMode {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_str() {
+            "front" => Ok(Front),
+            "back" => Ok(Back),
+            "all" => Ok(All),
+            _ => Err(InvalidArgument(format!(
+                "unsupported truncate mode {}, only support 'front', 'back', 'all'",
+                s
+            ))),
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct TruncateQueueParameter<'a> {
+    pub queue: Option<LogQueue>,
+    pub truncate_mode: TruncateMode,
+    pub raft_groups_ids: &'a [u64],
+}
+
+impl Default for TruncateMachine {
+    fn default() -> Self {
+        Self { items: vec![] }
+    }
+}
+
+impl ReplayMachine for TruncateMachine {
+    fn replay(&mut self, item_batch: LogItemBatch, _file_id: FileId) -> crate::Result<()> {
+        self.items.push(item_batch);
+        Ok(())
+    }
+
+    fn merge(&mut self, _rhs: Self, _queue: LogQueue) -> crate::Result<()> {
+        Ok(())
+    }
+
+    fn end<B: FileBuilder>(
+        &mut self,
+        path: &Path,
+        builder: Arc<B>,
+        truncate_params: &TruncateQueueParameter,
+    ) -> crate::Result<()> {
+        let origin_name = path.to_str().unwrap();
+        let truncate_file_name = origin_name.to_owned() + "_truncated";
+
+        let buf = PathBuf::from(truncate_file_name);
+        let fd = Arc::new(LogFd::create(buf.as_path())?);
+        let mut log_writer = build_file_writer(builder.as_ref(), buf.as_path(), fd, true)?;
+
+        let mut log_reader =
+            build_file_reader(builder.as_ref(), path, Arc::new(LogFd::open(path)?)).unwrap();
+
+        //write to file
+        self.flush_to_new_file::<B>(&mut log_writer, &mut log_reader, truncate_params)?;
+
+        //rename to original file name
+        fs::rename(buf.as_path(), path)?;
+        Ok(())
+    }
+}
+
+impl TruncateMachine {
+    fn flush_to_new_file<B: FileBuilder>(
+        &mut self,
+        log_writer: &mut LogFileWriter<B>,
+        log_reader: &mut LogFileReader<B>,
+        truncate_params: &TruncateQueueParameter,
+    ) -> crate::Result<()> {
+        let mut batch = LogBatch::default();
```

Comment:
Limit the size of one log batch.

Reply:
`LogBatch::default()` will make the size of log batch is `DEFAULT_LOG_ITEM_BATCH_CAP`(64),  Shall we need to change it explictly?

Reply:
I'm referring to the actual written size of one log batch. E.g. we limit the batch under `MAX_REWRITE_BATCH_BYTES` when doing purge rewrite.

Reply:
👌